File: C++/leetcode_questions/palindrome_linked_list.cpp
#include <bits/stdc++.h>
#include "../data_structures/ListNode.h"
























#pragma region more optimized Solution
// class Solution {
//     ListNode* reverseList(ListNode* head) {
//         ListNode* prev=nullptr;
//         ListNode* curr=head;
//         while(curr) {
//             ListNode* next=curr->next;
//             curr->next=prev;
//             prev=curr;
//             curr=next;
//         }
//         return prev;
//     }
// public:
//     bool isPalindrome(ListNode* head) {
//         int n=0;
//         ListNode* iter=head;
//         while(iter) iter=iter->next, n++;
//         if(n<2) return true;
//         // printf("-->%d\n", n);
//         iter=head;
//         for(int i=1;i<n/2;++i) iter=iter->next;
//         ListNode* iter2 = (n%2) ? iter->next->next : iter->next;
//         iter->next=nullptr;
//         // printf("iter2->val=%d\n",xiter2->val);
//         iter=reverseList(head);
//         while(iter) {
//             if(iter->val!=iter2->val) return false;
//             iter=iter->next;
//             iter2=iter2->next;
//         }
//         return true;
//     }
// };
#pragma endregion

File: python/leetcode_questions/palindromeLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from data_structures import ListNode

class Solution:
    
    '''
    1->2->3->2->1->null
    
    '''
    
    def isPalindrome(self, head: ListNode) -> bool:
        def reverseLst(head) -> ListNode:
            t1 = None
            t2 = head
            t3 = head.next
            while t3:
                t2.next = t1
                t1 = t2
                t2 = t3
                t3 = t3.next
            t2.next = t1
            return t2
        
        if not head.next: return True
        
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        secondHalf = None
        # length is odd
        if fast:
            secondHalf = slow.next
        # length is even
        else:
            secondHalf = slow
        # reverse the second half
        secondHalf = reverseLst(secondHalf)
        # sever the linked list
        slow.next = None
        
        
        while secondHalf and head:
            if head.val != secondHalf.val:
                return False
            head = head.next
            secondHalf = secondHalf.next
        return True
            
        
        
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            